src/main/java/org/usfirst/frc4048/common/apriltags/TCPServer.java

```diff
@@ -0,0 +1,120 @@
+package org.usfirst.frc4048.common.apriltags;
+
+import edu.wpi.first.wpilibj.DriverStation;
+
+import java.io.BufferedInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.LinkedTransferQueue;
+
+public abstract class TCPServer<T> extends Thread {
+
+    private final int port;
+    private ServerSocket serverSocket;
+    private Socket clientSocket;
+    private DataInputStream dataInputStream;
+    private BufferedInputStream bufferedInputStream;
+    private boolean running = false;
+    private final LinkedTransferQueue<T> readings = new LinkedTransferQueue<>();
+
+    /**
+     * @param port for network communication. Port must match between client and server (range 5800-5810 (rio limitations))
+     */
+    public TCPServer(int port) {
+        this.port = port;
+    }
+
+
+    @Override
+    public synchronized void start() {
+        try {
+            this.serverSocket = new ServerSocket(port);
+            this.clientSocket = null;
+        } catch (IOException e2) {
+            DriverStation.reportError("Could not start server on port " + port, true);
+            this.serverSocket = null;
+        }
+        running = true;
+        super.start();
+    }
+
+    @Override
+    public void run() {
+        while (running) {
+            if (serverSocket == null) {
+                exit();
```

Comment:
Returning from the `run` method would exit the thread, no need to call `exit`

Reply:
Wanted to call my `exit();` method, so the sockets would close before exiting the thread. Since it returns after calling `exit();` method, is the `this.join();` line necessary in the `exit();` method?

Reply:
No need to join once returning from the `run()`

Reply:
Unless you want to `join` on another thread for some reason, which is not the case for us.

Reply:
done